### ISSUE TITLE
Permission in PUBLIC role removes API permission from ADMIN role.

### ISSUE THREAD

**Issue by rodmaz** [apache/superset#24837]
:bug: Something strange is happening to Superset v2.1.0 regarding permissions.
Why adding a permission like can _read on dashboards_ for role **Public** removes this permission for role Admin?

#### How to reproduce the bug

1. Make sure **Public** role has NO permission.
2. Using `curl` and a token from an **Admin** user, query the API for available dashboards:

```[shell]
curl --location 'http://localhost:8088/api/v1/dashboard/' \
--header 'Authorization: Bearer xxxxx'
```

We should get a list of current available dashboards. (So far so good).
```
{
    "count": 10,
    "description_columns": {},
    "ids": [
        9,
        10,
        8,
...
```

4. Now add permission `can read on Dashboard` in the role **Public**.

<img width="495" alt="Screenshot 2023-07-28 at 17 28 06" src="https://github.com/apache/superset/assets/55300/813c3ae5-e045-4418-92f1-ff136fdd2179">
 
5. Perform the same query as step 2 and (surprisingly!) we see no more Dashboards in the API.

```
{
  "count": 0, 
  "description_columns": {}, 
  "ids": [], 
  "label_columns"
  ...
```

### Expected results

Adding permission in **Public** role should not affect users using the API with role **Admin**.

### Actual results

Adding permission in role **Public** REMOVES permission from users with role **Admin**.

### Environment

- superset version: `2.1.0` and `2.1.1rc2`
- python version: `3.8.17`
- node.js version: `v16.20.0`

- [x] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [x] I have reproduced the issue with at least the latest released version of superset.
- [x] I have checked the issue tracker for the same issue and I haven't found one similar.

### Additional context

**Public** is the role defined for guest users:

```[python]
# Embedded config options
GUEST_ROLE_NAME = "Public"
GUEST_TOKEN_JWT_SECRET = "test-guest-secret-change-me"
GUEST_TOKEN_JWT_ALGO = "HS256"
GUEST_TOKEN_HEADER_NAME = "X-GuestToken"
GUEST_TOKEN_JWT_EXP_SECONDS = 300  # 5 minutes
# Guest token audience for the embedded superset, either string or callable
GUEST_TOKEN_JWT_AUDIENCE: Optional[Union[Callable[[], str], str]] = None
```

**Comment by mdeshmu** [comment]
It is not recommended to modify default roles. 
You should clone the default roles and make modifications to them. 
Are you able to replicate this in cloned roles?

**Comment by rodmaz** [comment]
We are only adding permissions to the **Public** role since it is a requirement when we embed dashboards.
That’s the overall approach documented. We just didn’t expect it to break the API.
**Public** role is our guest role.

**Comment by rodmaz** [comment]
Should we create a new role for guest users instead?

**Comment by rodmaz** [comment]
We managed to fix this by using another custom role for the `GUEST_ROLE_NAME` and leaving `PUBLIC` alone.
Still odd, unexpected behavior but this works fine.

**Comment by ahmed-adly-khalil** [comment]
i'm having same issue, why adding permission to the public rule removes it from the admin and on the api level only? !!

**Comment by nytai** [comment]
@ahmed-adly-khalil Embedded doesn't have the use the Public role, you can configure it to use whichever role you want with `GUEST_ROLE_NAME` as @rodmaz has done. This definitely seems like an issue, but there are workaround. 

@dpgaspar any thoughts on what might be happening here?

**Comment by ahmed-adly-khalil** [comment]
@nytai Yeah `GUEST_ROLE_NAME` fixed the problem for me and i can do both dashboard embedding and use the API, thanks a lot. I still think this is a bug and it's great to be fixed since this behaviour is mostly unexpected

**Comment by tcape** [comment]
@rodmaz @ahmed-adly-khalil Could either of you provide your `config.py` and `superset_config.py` values you have set? I cannot get embedding and API to both work. It's always either one or the other.

**Comment by ahmed-adly-khalil** [comment]
Just use a different role for embracing not public

**Comment by rusackas** [comment]
Just checking in to see if anyone can recapitulate the issue here. I'm tempted to close it as stale, but it's probably gone stale because it's hard to tell if there's a real bug here, or if this is a question of how to configure roles for embedding (in which case, it should be a Discussion, not an Issue). Thoughts?

**Comment by ss-ravi** [comment]
@rusackas Adding permissions to the Public role makes the api requests using the Admin role stop working, so it does look like a bug. https://github.com/apache/superset/issues/26470

**Comment by rodmaz** [comment]
This is a well described and rather serious bug I would say, when users add permissions for role **Public** removes permissions for role **Admin**.

**Comment by rusackas** [comment]
@rodmaz a serious bug on 2.1.0, or have you upgraded since then?

**Comment by rodmaz** [comment]
We haven't upgraded yet but plan to do so soon.
I would say, if this is not a bug it should be definitely better documented. Adding permissions in one role ending up removing permissions from another is not something users really expect TBH.

**Comment by rusackas** [comment]
Agreed... it's definitely a bug if it's happening in supported versions. That just needs to be validated.

**Comment by ss-ravi** [comment]
@rusackas I haven't tried it on the 4.0.0 version, but it does happen on 3.0.2.

**Comment by rusackas** [comment]
Good to know... but we currently support 3.1.x and 4.0.x for the time being. If anyone can validate using those versions, that would be even better still.

**Comment by aibunny** [comment]
For superset version 3.1.x I didn't experience this issue but for 4.0.x I am experiencing this issue.

**Comment by kleky** [comment]
I can confirm this occurs on 4.0.0 and is causing us serious pain. We sync DB schema from Cube Cloud to Superset, which uses the REST API. The workaround by changing the GUEST_ROLE_NAME worked sometimes, but not on other instances and it makes no sense why.

**Comment by FynnMazurkiewicz** [comment]
TLDR: **Patch available!**

This bug is still present in version 4.0.1. The workaround suggested by @lbuchli did not work for me either.

**Bug summary**
Whenever you set the "can_read" permission of "Dashboard" as part of the public permissions, the API of that endpoint will start ignoring the JWT entirely and always assume that the public user makes the request. In fact, this bug not only occurs for this endpoint, but for all endpoints. As soon as any item is part of the public read permissions, the JWT is ignored. This breaks the JWT authorization, as the API ignores any given token and always assumes public user permissions. For example, the dashboard endpoint will always return only dashboards available to the public user, even if you supply a JWT Bearer token of a user that has access to more dashboards (such as the admin role).

**Root cause**
I have completed a root cause analysis of this bug. It is caused by an issue in FlaskAppBuilder, the underlying framework also built by the lovely @dpgaspar. [In the security decorator `_protect()`](https://github.com/dpgaspar/Flask-AppBuilder/blob/master/flask_appbuilder/security/decorators.py#L98), that is called to set the authorization context on any endpoint, there is a special condition set for items that are considered public (meaning the public permissions can read them). I'll quote the relevant parts for discussion:

```python
flask_appbuilder/security/decorators.py:98

# Check if the resource is public
if current_app.appbuilder.sm.is_item_public(permission_str, class_permission_name):
    return f(self, *args, **kwargs)

# if no browser login then verify JWT
if not (self.allow_browser_login or allow_browser_login):
    verify_jwt_in_request()

# Verify resource access
if current_app.appbuilder.sm.has_access(permission_str, class_permission_name):
    return f(self, *args, **kwargs)
```
In this decorator, it is first checked whether the item is public and if it is, the endpoint function is immediately called - with no further JWT token handling. You can see the function call to `verify_jwt_in_request()` is only called when the public condition check fails.

I am not sure if this can be considered a bug in FAB itself, so I won't open a bug issue there. I do see however, that this might cause issues a lot and I highly question this order is correct. @dpgaspar What do you think?

**Patch**
Until there is a proper fix implemented by @dpgaspar or any other maintainer, its possible to hack around this by attempting a JWT login by overwriting `is_item_public` in the SecurityManager and adding a call to `verify_jwt_in_request` there.
```python
class MySecurityManager(SupersetSecurityManager):
    ...
    def is_item_public(self, permission_name, view_name):
        verify_jwt_in_request(optional=True) # Attempt to parse any existing JWT and fail silently
        return super().is_item_public(permission_name, view_name)
    ...
```
Thanks to my employer https://www.dksr.city/en/ for donating A LOT of time to work on this and being able to donate my findings back to the open source community.

**Comment by zachoooo** [comment]
@FynnMazurkiewicz This explains so much! I've been really struggling with this issue. I wonder why this issue doesn't occur if you set the session cookie. The one provided by CSRF token endpoint doesn't work, but the one provided by actually logging into Superset in browser and then inspecting the network and copying will work properly without any changes.

**Comment by rusackas** [comment]
This has been silent for a while, so it hit my "stale issues" radar. Does anyone (including @dpgaspar) have an update here as to whether or not this has been patches in FAB? Not sure if we should keep this open or not on this repo. Maybe there's an issue on the FAB repo we can link to here?

**Comment by lautaro79** [comment]
In my humble opinion @FynnMazurkiewicz  found a great workaround and it should be added directly to the SupersetSecurityManager to avoid everyone implementing the same thing in a CustomSecurityManager.

The solution I am referring to is: [Patch](https://github.com/apache/superset/issues/24837#issuecomment-2501991590)

**Comment by 7frank** [comment]
It took a day to get back to this topic. I can confirm that https://github.com/apache/superset/issues/24837#issuecomment-2501991590 addresses the issue for me. Although I agree with @lautaro79 that this could reduce a lot of head scratch when added by default.

**Comment by rusackas** [comment]
Since the comment addressed the issue, this is otherwise quite stale, we have a new Public default role now, and this issue was reported on 2.0, I think I'm going to take the liberty of closing this one out.

If anyone is still experiencing an issue in this department, please feel free to regroup and post it as a new issue so it's easier to grok and address.
